server/WFCORE-4251_list_modules_in_deployment_classpath.adoc

```diff
@@ -0,0 +1,170 @@
+      "result" => {
+          "system-dependencies" => [
+              {"name" => "com.fasterxml.jackson.datatype.jackson-datatype-jdk8"},
+              {"name" => "com.fasterxml.jackson.datatype.jackson-datatype-jsr310"},
+              {"name" => "ibm.jdk"},
+              {"name" => "io.jaegertracing.jaeger"},
+              {"name" => "io.opentracing.contrib.opentracing-tracerresolver"},
+              ...
+          ],
+          "local-dependencies" => [
+              {"name" => "deployment.ejb-in-ear.ear.ejb-in-ear-ejb.jar"},
+               ...
+          ],
+          "user-dependencies" => [
+              {"name" => "com.fasterxml.jackson.datatype.jackson-datatype-jdk8"},
+              {"name" => "org.hibernate:4.1"},
+               ...
+          ]
+      }
+  }
+
+And the detailed output could be:
+
+  [standalone@localhost:9990 /] /deployment=sample-ear-1.0.ear:list-modules(verbose=true)
+  {
+      "outcome" => "success",
+      "result" => {
+          "system-dependencies" => [
+              {
+                  "name" => "com.fasterxml.jackson.datatype.jackson-datatype-jdk8",
+                  "optional" => true,
+                  "export" => false,
+                  "import-services" => true
+              },
+              {
+                  "name" => "com.fasterxml.jackson.datatype.jackson-datatype-jsr310",
+                  "optional" => true,
+                  "export" => false,
+                  "import-services" => true
+              },
+              ...
+          ],
+          "local-dependencies" => [
+              {
+                "name" => "deployment.ejb-in-ear.ear.ejb-in-ear-ejb.jar",
+                "optional" => false,
+                "export" => false,
+                "import-services" => true
+              },
+              ...
+          ],
+          "user-dependencies" => [
+              {
+                  "name" => "com.fasterxml.jackson.datatype.jackson-datatype-jdk8",
+                  "optional" => false,
+                  "export" => false,
+                  "import-services" => false
+              },
+              {
+                  "name" => "org.hibernate:4.1",
+                  "optional" => false,
+                  "export" => false,
+                  "import-services" => false
+              },
+              ...
+          ]
+      }
+  }
+
+The information will be presented in three different categories:
+
+* system-dependencies: These are the dependencies added implicitly by the server.
+* local-dependencies: These are dependencies on other parts of the deployment.
+* user-dependencies: These are the dependencies defined by the user via manifest file or deployment-structure.xml.
+
+
+For each category, the meaning of each attribute is:
+
+* name: The module name and, if the slot name is not the default "main" slot, the slot name is concatenated after ":" character separator.
+* optional: If the dependency was added as an optional dependency.
```

Comment:
I interpret 'optional' as meaning the module might not be present, but either the user (for user dependencies) or the server (for system and local) recorded dependency info in case it is present. But then, if optional=true, can we reliably report whether the dependency is actually present?  If yes, that's something to consider for any future iteration on this feature. I don't think it's necessary for this first iteration though

Reply:
> I interpret 'optional' as meaning the module might not be present

Correct.

> But then, if optional=true, can we reliably report whether the dependency is actually present?

Yes, we can detect if the dependency, being optional, is not included in the server distribution, and then, somehow inform about this situation in the report. I can see some cases that have to be handled, e.g. a deployment depends optionally in another deployment and that second deployment is undeploy, we must deal with these cases as well to reflect correctly those runtime changes. But, yes, we could deal with it. It is left for a second iteration in case of be requested.